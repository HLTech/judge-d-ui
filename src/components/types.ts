import { SimulationLinkDatum, SimulationNodeDatum } from 'd3-force';
import { BaseType, Selection } from 'd3-selection';

export interface Network {
    nodes: DependencyNode[];
    links: DependencyLink[];
    detailsNodes: TreeNode[];
}

export interface DependencyNode extends SimulationNodeDatum {
    name: string;
    version: string;
    isProvider: boolean;
    isConsumer: boolean;
    width?: number;
    height?: number;
    links: Array<DependencyNode>;
    level: number;
}

export interface DependencyLink extends SimulationLinkDatum<DependencyNode> {
    source: DependencyNode;
    target: DependencyNode;
    type: string;
}

export interface TreeNode {
    name: string;
    consumers: TreeNode[];
    providers: TreeNode[];
}

export interface Service {
    name: string;
    version: string;
    capabilities: ServiceCommunication;
    expectations: Expectations;
}

export interface Expectations {
    [key: string]: ServiceCommunication;
}

export interface ServiceCommunication {
    [key: string]: {
        value: object;
        mimeType: string;
    };
}

export type NodeSelection<T extends BaseType> = Selection<T, any, Element, HTMLElement>;
