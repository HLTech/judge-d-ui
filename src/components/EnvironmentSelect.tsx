import * as React from 'react';
import { Dropdown, DropdownProps } from 'semantic-ui-react';
import { css } from 'emotion';
import { useMemo } from 'react';

export interface EnvironmentSelectProps {
    env: string;
    disabled: boolean;
    options: string[];

    onEnvironmentChange(env: string): void;
}

export const EnvironmentSelect: React.FC<EnvironmentSelectProps> = ({ env, options, disabled, onEnvironmentChange }) => {
    const dropdownOptions = useMemo(
        () =>
            options.map((type: string) => ({
                text: type,
                value: type,
            })),
        [options]
    );

    const handleEnvironmentChange = (event: React.SyntheticEvent<HTMLElement>, data: DropdownProps) => {
        if (data.value) {
            onEnvironmentChange(data.value as string);
        }
    };

    return (
        <div className={dropdownContainerCls}>
            <span>Select Environment</span>

            <Dropdown
                className={dropdownCls}
                data-test="environment-select"
                fluid={true}
                multiple={false}
                selection={true}
                options={dropdownOptions}
                onChange={handleEnvironmentChange}
                value={env}
                disabled={disabled}
            />
        </div>
    );
};

const dropdownContainerCls = css({
    label: 'env-select-container',
    minWidth: 100,
});

const dropdownCls = css({
    label: 'env-select',
    minWidth: 100,
    display: 'inline-block',
});
