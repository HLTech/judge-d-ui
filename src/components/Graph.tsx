import * as React from 'react';
import { useRef } from 'react';
import { css } from 'emotion';
import { useEffect } from 'react';
import { draw } from '../utils/Draw';
import * as d3 from 'd3';
import { DependencyLink, DependencyNode } from './types';

export interface GraphProps {
    nodes: DependencyNode[];
    links: DependencyLink[];
}

export const Graph = React.memo<GraphProps>(({ nodes, links }) => {
    const dependencyGraphDiv = useRef<HTMLDivElement>(null);

    useEffect(() => {
        d3.select(dependencyGraphDiv.current)
            .select('*')
            .remove();
        if (dependencyGraphDiv.current !== null && nodes.length > 0) {
            draw({ nodes, links }, dependencyGraphDiv.current);
        }
    }, [nodes, links]);

    return <div ref={dependencyGraphDiv} className={graphCls} id={'dependency-graph-container'} />;
});

const graphCls = css({
    display: 'flex',
    flexDirection: 'column',
    overflow: 'hidden',
    height: '80vh',
    '& > svg': {
        flex: 1,
        overflow: 'hidden',
    },
});
