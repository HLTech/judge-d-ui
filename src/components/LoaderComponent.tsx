import * as React from 'react';
import { Loader } from 'semantic-ui-react';
import { css } from 'emotion';

export const LoaderComponent: React.FC = () => {
    return (
        <div className={loaderCls} data-test="loader">
            <Loader active inline="centered" size="large" />
        </div>
    );
};

const loaderCls = css({
    height: '80vh',
    width: '100vw',
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'space-evenly',
});
