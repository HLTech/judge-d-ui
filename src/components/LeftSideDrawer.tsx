import React from 'react';
import { EnvironmentSelect } from './EnvironmentSelect';
import { css } from 'emotion';
import { CheckboxProps, Radio } from 'semantic-ui-react';

interface LeftSideDrawerProps {
    selectedEnvironment: string;
    environmentOptions: string[];
    areControlsDisabled: boolean;
    areOnlyConnectedNodesShown: boolean;

    closeMenu(): void;
    handleViewSwitchChange(e: React.FormEvent<HTMLInputElement>, data: CheckboxProps): void;
    onEnvironmentChange(env: string): void;
}

export const LeftSideDrawer: React.FC<LeftSideDrawerProps> = ({
    selectedEnvironment,
    environmentOptions,
    areControlsDisabled,
    onEnvironmentChange,
    closeMenu,
    areOnlyConnectedNodesShown,
    handleViewSwitchChange,
}) => (
    <div className={backgroundCls}>
        <div className={containerCls}>
            <h1 className={headerCls}>judge-d</h1>
            <p className={descriptionCls}>Contract testing tool for microservices architecture</p>
            <EnvironmentSelect
                env={selectedEnvironment}
                disabled={areControlsDisabled}
                options={environmentOptions}
                onEnvironmentChange={onEnvironmentChange}
            />
            {process.env.NODE_ENV === 'development' && (
                <Radio onChange={handleViewSwitchChange} toggle checked={areOnlyConnectedNodesShown} />
            )}
        </div>
        <div className={shadowCls} onClick={() => closeMenu()} />
    </div>
);

const backgroundCls = css({
    position: 'fixed',
    display: 'grid',
    gridTemplateColumns: '320px 1fr',
    zIndex: 10,
    left: 0,
    top: 0,
    right: 0,
    bottom: 0,
});

const headerCls = css({
    fontSize: 40,
});

const descriptionCls = css({
    fontSize: 13,
});

const containerCls = css({
    display: 'grid',
    gridTemplateRows: '60px 60px auto',
    borderTop: '15px solid #E5E5E6',
    backgroundColor: '#ffffff',
    padding: '32px 38px',
    boxShadow: '0px 0px 60px #00000010',
});

const shadowCls = css({
    backgroundColor: '#00000040',
});
