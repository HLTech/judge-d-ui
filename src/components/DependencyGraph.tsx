import * as React from 'react';
import { useEffect, useMemo, useState } from 'react';
import { LoaderComponent } from './LoaderComponent';
import { DependencyNode, Network, Service } from './types';
import { createNetworkFromServices, filterConnectedNodes } from '../utils/helpers/MappingHelpers';
import { Button, CheckboxProps, Icon } from 'semantic-ui-react';
import { Graph } from './Graph';
import { getServicesRequest, getEnvironmentsRequest } from '../api/api';
import 'semantic-ui-css/semantic.css';
import { css } from 'emotion';
import { LeftSideDrawer } from './LeftSideDrawer';

export const DependencyGraph: React.FC = () => {
    const [nodes, setNodes] = useState<DependencyNode[]>([]);
    const [isPending, setIsPending] = useState<boolean>(true);
    const [services, setServices] = useState<Service[]>([]);
    const [env, setEnv] = useState<string>('');
    const [environments, setEnvironments] = useState<string[]>([]);
    const [shouldShowAllNodes, setShouldShowAllNodes] = useState<boolean>(false);
    const [isMenuOpen, setIsMenuOpen] = useState(false);

    const graphNetwork: Network = useMemo(() => createNetworkFromServices(services), [services]);
    const onlyNodesWithContracts = useMemo(() => filterConnectedNodes(graphNetwork), [graphNetwork]);

    useEffect(() => {
        setIsPending(true);
        getEnvironmentsRequest()
            .then((environments: string[]) => {
                setEnvironments(environments);
                setEnv(environments[0]);
                setIsPending(false);
            })
            .catch((error: Error) => {
                console.log(error);
                setIsPending(false);
            });
    }, []);

    useEffect(() => {
        if (env) {
            setIsPending(true);
            getServicesRequest(env)
                .then((data: Service[]) => {
                    setServices(data);
                    setIsPending(false);
                })
                .catch((error: Error) => {
                    console.log(error);
                    setIsPending(false);
                });
        }
    }, [env]);

    const onViewSwitchChange = (e: React.FormEvent<HTMLInputElement>, data: CheckboxProps) => setShouldShowAllNodes(Boolean(data.checked));

    useEffect(() => {
        if (shouldShowAllNodes) {
            setNodes(graphNetwork.nodes);
        } else {
            setNodes(onlyNodesWithContracts);
        }
    }, [shouldShowAllNodes, onlyNodesWithContracts, graphNetwork]);

    return (
        <>
            <div className={drawerOpenButtonCls}>
                <Button icon onClick={() => setIsMenuOpen(true)}>
                    <Icon name={'bars'} size={'large'} />
                </Button>
            </div>
            {isPending && <LoaderComponent />}

            <Graph network={{ ...graphNetwork, nodes }} />
            {isMenuOpen && (
                <LeftSideDrawer
                    environmentOptions={environments}
                    selectedEnvironment={env}
                    areControlsDisabled={isPending}
                    onEnvironmentChange={setEnv}
                    onBackgroundClick={() => setIsMenuOpen(false)}
                    shouldShowAllNodes={shouldShowAllNodes}
                    onViewSwitchChange={onViewSwitchChange}
                />
            )}
        </>
    );
};

const drawerOpenButtonCls = css({
    label: 'drawer-open-button',
    position: 'fixed',
    margin: 20,
    zIndex: 5,
});
