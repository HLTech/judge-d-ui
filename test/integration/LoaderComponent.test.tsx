import * as React from 'react';
import { render } from '@testing-library/react';
import { LoaderComponent } from '../../src/components/LoaderComponent';

describe('LoaderComponent', () => {
    it('should render without error', () => {
        const { container } = render(<LoaderComponent />);
        expect(container.firstChild).toMatchSnapshot();
    });
});
