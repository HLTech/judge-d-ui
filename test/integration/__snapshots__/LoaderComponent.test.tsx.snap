// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoaderComponent should render without error 1`] = `
<div
  class="css-pmoqd7-loader"
  data-test="loader"
>
  <div
    class="ui large active centered inline loader"
  />
</div>
`;
